models/aeif_cbvg_2010.cpp

```diff
@@ -0,0 +1,657 @@
+
+  V_.V_peak_ = P_.V_peak_;
+
+  V_.refractory_counts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  // since t_ref_ >= 0, this can only fail in error
+  assert( V_.refractory_counts_ >= 0 );
+
+  // implementation of the clamping after a spike
+  V_.clamp_counts_ = Time( Time::ms( P_.t_clamp_ ) ).get_steps();
+  // since t_clamp_ >= 0, this can only fail in error
+  assert( V_.clamp_counts_ >= 0 );
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::aeif_cbvg_2010::update( const Time& origin,
+  const long from,
+  const long to )
+{
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+  assert( State_::V_M == 0 );
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+    double t = 0.0;
+
+    // numerical integration with adaptive step size control:
+    // ------------------------------------------------------
+    // gsl_odeiv_evolve_apply performs only a single numerical
+    // integration step, starting from t and bounded by step;
+    // the while-loop ensures integration over the whole simulation
+    // step (0, step] if more than one integration step is needed due
+    // to a small integration step size;
+    // note that (t+IntegrationStep > step) leads to integration over
+    // (t, step] and afterwards setting t to step, but it does not
+    // enforce setting IntegrationStep to step-t
+    while ( t < B_.step_ )
+    {
+      const int status = gsl_odeiv_evolve_apply( B_.e_,
+        B_.c_,
+        B_.s_,
+        &B_.sys_,             // system of ODE
+        &t,                   // from t
+        B_.step_,             // to t <= step
+        &B_.IntegrationStep_, // integration step size
+        S_.y_ );              // neuronal state
+
+      if ( status != GSL_SUCCESS )
+      {
+        throw GSLSolverFailure( get_name(), status );
+      }
+      // check for unreasonable values; we allow V_M to explode
+      if ( S_.y_[ State_::V_M ] < -1e3 || S_.y_[ State_::W ] < -1e6
+        || S_.y_[ State_::W ] > 1e6 )
+      {
+        throw NumericalInstability( get_name() );
+      }
+
+      // spikes are handled inside the while-loop
+      // due to spike-driven adaptation
+      if ( S_.r_ == 0 && S_.clamp_r_ == 0 )
+      {
+        // neuron not refractory
+        S_.y_[ State_::V_M ] =
+          S_.y_[ State_::V_M ] + B_.spikes_.get_value( lag );
+      }
+      else // neuron is absolute refractory
+      {
+        B_.spikes_.get_value( lag ); // clear buffer entry, ignore spike
+      }
+
+      // set the right threshold depending on Delta_T
+      if ( P_.Delta_T == 0. )
+      {
+        V_.V_peak_ = S_.y_[ State_::V_T ]; // same as IAF dynamics for spikes if
+                                           // Delta_T == 0.
+      }
+
+      if ( S_.y_[ State_::V_M ] >= V_.V_peak_ && S_.clamp_r_ == 0 )
```

Comment:
I don't know the clopath model and I unfortunately do not have time to check it right now, but this section and the ``elif`` look very weird to me, I take it that ``S_.clamp_r_ == 1`` **must** immediately follow a spike, otherwise there is something weird going on, is that correct? (if everything is fine, then I guess I'll just come back to it when I have time to actually check the model)

Reply:
`clamp_r` is a counter that is counting down the time steps left where the membrane potential is clamed to `V_clamp`. So `S_clamp_r == 1` is the case that occurs at the end of the clamping. The additional check `&& S_.clamp_r_ == 0` in the original `if` is just to ensure that `V_M` is not just greater than `V_peak` because it is clamped at the moment (we actually allow `V_clamp > V_peak`).
Can this explanation resolve the confusion? I added a comment explaining `clamp_r` in the code.